pkg/operator/observe_config.go

```diff
@@ -149,13 +154,172 @@ func observeInternalRegistryHostname(listers Listers, observedConfig map[string]
+	}
+	if err = observeField(observedConfig, build.Spec.BuildDefaults.GitHTTPSProxy, "build.buildDefaults.gitHTTPSProxy"); err != nil {
+		return nil, fmt.Errorf("failed to observe %s: %v", "build.buildDefaults.gitHTTPSProxy", err)
+	}
+	if err = observeField(observedConfig, build.Spec.BuildDefaults.GitNoProxy, "build.buildDefaults.gitNoProxy"); err != nil {
+		return nil, fmt.Errorf("failed to observe %s: %v", "build.buildDefaults.gitNoProxy", err)
+	}
+	if err = observeField(observedConfig, build.Spec.BuildDefaults.Env, "build.buildDefaults.env"); err != nil {
+		return nil, fmt.Errorf("failed to observe %s: %v", "build.buildDefaults.env", err)
+	}
+	if err = observeField(observedConfig, build.Spec.BuildDefaults.ImageLabels, "build.buildDefaults.imageLabels"); err != nil {
+		return nil, fmt.Errorf("failed to observe %s: %v", "build.buildDefaults.imageLabels", err)
+	}
+
+	// set build overrides
+	if err = observeField(observedConfig, build.Spec.BuildOverrides.ImageLabels, "build.buildOverrides.imageLabels"); err != nil {
+		return nil, fmt.Errorf("failed to observe %s: %v", "build.buildOverrides.imageLabels", err)
+	}
+	nodeSelector := build.Spec.BuildOverrides.NodeSelector
+	if err = observeField(observedConfig, nodeSelector.MatchLabels, "build.buildOverrides.nodeSelector"); err != nil {
+		return nil, fmt.Errorf("failed to observe %s: %v", "build.buildOverrides.nodeSelector", err)
+	}
+	// Control plane config does not support MatchExpressions yet
+	if len(nodeSelector.MatchExpressions) > 0 {
+		return nil, fmt.Errorf("config.Build: %s is not supported", "buildOverrides.nodeSelector.matchExpressions")
+	}
+	if err = observeField(observedConfig, build.Spec.BuildOverrides.Tolerations, "build.buildOverrides.tolerations"); err != nil {
+		return nil, fmt.Errorf("failed to observe %s: %v", "build.buildOverrides.tolerations", err)
+	}
+
+	// TODO: 1) generate CA bundle from ConfigMapRef
+	//       2) write CA bundle to a ConfigMap in the controller-manager's namespace
+	//       3) wire in logic to force a rollout if CA bundle changes
+	// additionalCA := build.Spec.AdditionalTrustedCA
+	// if len(additionalCA.Name) > 0 {
+	// 	unstructured.SetNestedField(observedConfig, "/var/run/openshift.io/config/certs/additional-ca.crt", "build", "additionalTrustedCA")
+	// }
+	return observedConfig, nil
+}
+
+// observeField sets the nested fieldName's value in the provided observedConfig.
+// If the provided value is nil or empty, no value is set.
+func observeField(observedConfig map[string]interface{}, val interface{}, fieldName string) error {
+	nestedFields := strings.Split(fieldName, ".")
+	if val == nil {
+		return nil
+	}
+	var err error
+	switch v := val.(type) {
+	case int64, bool:
+		err = unstructured.SetNestedField(observedConfig, v, nestedFields...)
+	case string:
+		if len(v) > 0 {
+			err = unstructured.SetNestedField(observedConfig, v, nestedFields...)
+		}
```

Comment:
i'm not sure how i feel about baking in the assumption that we'd never want/need to set a field to empty string, but I don't think I'm a fan.  I think i prefer that the caller makes the decision about whether to set the field or not, in that case.

Reply:
You can make the same argument for all types. Is it worth keeping this function by adding a parameter, or a separate function def?

Reply:
I intended to make the argument for all types, sorry that wasn't clear :)

I don't have a strong preference between adding an "skipEmpty" bool and making this function always invoke the set and relying on the caller to not call it if they don't want it set.

I think i have a slight preference towards the latter in case the decision making gets even more subtle (e.g. different behavior for a nil pointer to a slice vs an empty slice)